sky/skylet/providers/gcp/config.py

```diff
@@ -504,24 +505,58 @@ def _check_firewall_rules(vpc_name, config, compute):
         return False
     effective_rules = response["firewalls"]
 
-    def _get_refined_rule(rule):
-        KEY_TO_COMPARE = {"sourceRanges", "allowed", "direction"}
-        refined_rule = {}
-        for k in KEY_TO_COMPARE:
-            if k not in rule:
-                continue
-            if k == "allowed":
-                refined_rule[k] = sorted(rule[k], key=lambda x: x["IPProtocol"])
-            else:
-                refined_rule[k] = rule[k]
-        return refined_rule
-
-    required_rules = list(map(_get_refined_rule, required_rules))
-    effective_rules = list(map(_get_refined_rule, effective_rules))
-
-    for rule in required_rules:
-        if rule not in effective_rules:
+    def _merge_and_refine_rule(rules):
+        # Example of source2rules: Dict[(direction, sourceRanges) -> Dict(protocol -> Set[ports])]
+        #   {("INGRESS", "0.0.0.0/0"): {"tcp": {80, 443}, "udp": {53}}}
```

Comment:
How about adding a docstr for this function as it is pretty long?
```
"""Returns the reformatted rules from the firewall rules

The function translates the firewall rules fetched from the cloud provider
to the easier to compare format

Firewall rules from the cloud:
[
    {
        ...
        "direction": "INGRESS",
        "allowed": [
            {'IPProtocol': 'tcp', 'ports': ['0-65535']},
            {'IPProtocol': 'udp', 'ports': ['0-65535']},
        ],
        "sourceRanges": ["10.128.0.0/9"],
    },
    {
        ...
        "direction": "INGRESS",
        "allowed": [{
            "IPProtocol": "tcp",
            "ports": ["22"],
        }],
        "sourceRanges": ["0.0.0.0/0"],
    },
]

Returned format (Please correct this)
{("INGRESS", "0.0.0.0/0"): {"tcp": {80, 443}, "udp": {53}}}

Returns:
    source2rules: Dict[(direction, sourceRanges) -> Dict(protocol -> Set[ports])]
        Example {("INGRESS", "10.128.0.0/9"): {"tcp": {80, 443}, "udp": {53}}}
"""
```

Reply:
Great idea. I just added the comments below the function.